File: content/problems/2349-design-a-number-container-system/solutions/bidirectional-maps-with-ordered-indices/solution.cpp
class NumberContainers {
    unordered_map<int,int> indexToNumber;
    unordered_map<int,set<int>> numberToIndices;
    public:NumberContainers(){
    }
    void change(int index,int number){
        if(indexToNumber.count(index))numberToIndices[indexToNumber[index]].erase(index);
        indexToNumber[index]=number;
        numberToIndices[number].insert(index);
    }
    int find(int number){
        return numberToIndices[number].empty()?-1:*numberToIndices[number].begin();
    }
};

File: content/problems/2349-design-a-number-container-system/solutions/bidirectional-maps-with-ordered-indices/Solution.kt
class NumberContainers {
    private val indexToNumber=HashMap<Int,Int>();
    private val numberToIndices=HashMap<Int,java.util.TreeSet<Int>>();
    fun change(index:Int,number:Int){
        indexToNumber[index]?.let{
            numberToIndices[it]!!.remove(index)
        };
        indexToNumber[index]=number;
        numberToIndices.getOrPut(number){
            java.util.TreeSet()
        }.add(index)
    }
    fun find(number:Int):Int{
        return numberToIndices[number]?.firstOrNull()?:-1
    }
}
